doc/fundamentals/spouts.md

```diff
@@ -0,0 +1,287 @@
+	out, err := os.OpenFile("/pfs/out", os.O_WRONLY, 0644)
+	if err != nil {
+		panic(err)
+	}
+	defer out.Close()
+```
+
+### Write the outer file loop
+Here we open a tar stream into the directory we opened above,
+so that Pachyderm can place files into the output repo.
+For clarity's sake,
+we'll omit the message-processing loop inside this file loop.
+The `err` variable is used in the message-processing loop for errors reading from the stream and writing to the directory.
+The stream is opened at the top of the loop and should be closed at the bottom.
+In this case, 
+it'll be closed after a message is processed.
+The commit is finished after the file stream closes.
+(The `defer tw.Close()` is wrapped in an anonymous function to control when it gets run; 
+it'll run after all the code in the anonymous function is finished,
+but before it returns.)
+Think of the `tw.Close()` as a `FinishCommit()`.
+
+```Go
+	// this is the file loop
+	for {
+		if err := func() error {
+			tw := tar.NewWriter(out)
+			defer tw.Close()
+			// this is the message loop
+			for {
+
+				// ...omitted
+
+			}
+		}(); err != nil {
+			panic(err)
+		}
+	}
+```
+
+### Create the message processing loop
+
+If you have trouble following this Go code,
+just read the text to get an idea of what you need to do.
+
+First,
+we read a message from our Kafka queue.
+Note the use of a 5-second backgrounded timeout on the read.
+That's so if the data source,
+Kafka,
+hangs for some reason, 
+the spout itself doesn't hang,
+but gives a hopefully useful error message in the logs after crashing.
+
+```Go
+			// this is the message loop
+			for {
+				// read a message
+				ctx, cancel := context.WithTimeout(context.Background(), 5*time.Second)
+				defer cancel()
+				m, err := reader.ReadMessage(ctx)
+				if err != nil {
+					return err
+				}
+```
+
+Then we create a filename
+and write it, 
+along with the file size
+in a file header, 
+to the `tar` stream
+we opened at the beginning
+of the file loop.
+This tar stream will be used by Pachyderm
+to create the file
+in the output repo.
+```Go
+				// give it a unique name
+				name := topic + time.Now().Format(time.RFC3339Nano)
+				// write the header
+				for err = tw.WriteHeader(&tar.Header{
+					Name: name,
+					Mode: 0600,
+					Size: int64(len(m.Value)),
+				}); err != nil; {
+					if !strings.Contains(err.Error(), "broken pipe") {
+						return err
+					}
+					// if there's a broken pipe, just give it some time to get ready for the next message
+					time.Sleep(5 * time.Millisecond)
+				}
+```
+
+Then we write the actual message as the contents of the file.
+Note the use of a timeout in case the named pipe is broken.
+The reason for this is that the other end of the spout
+(Pachyderm's code)
+has closed the named pipe at the end of the previous read. 
+Pachyderm may not have reopened the named pipe yet,
+so our code should back off if it gets an error writing to the named pipe.
```

Comment:
i know it says this if you read it carefully, but i think it should be more clear here that it should only back off if it gets a `broken pipe` error, idk